spring-kafka/src/test/java/org/springframework/kafka/listener/ConsumerRebalanceProtocolTests.java

```diff
@@ -0,0 +1,174 @@
+	}
+
+	private void testRebalance(ConsumerFactory<String, String> consumerFactory, String groupId, boolean isNewProtocol)
+			throws Exception {
+		AtomicReference<Consumer<?, ?>> consumerRef = new AtomicReference<>();
+		CountDownLatch revokedBeforeCommitLatch = new CountDownLatch(1);
+		CountDownLatch revokedAfterCommitLatch = new CountDownLatch(1);
+		CountDownLatch assignedLatch = new CountDownLatch(1);
+
+		Set<TopicPartition> assignedPartitionSet = new HashSet<>();
+
+		ConsumerAwareRebalanceListener listener = new ConsumerAwareRebalanceListener() {
+			@Override
+			public void onPartitionsRevokedBeforeCommit(Consumer<?, ?> consumer, Collection<TopicPartition> partitions) {
+				consumerRef.set(consumer);
+				revokedBeforeCommitLatch.countDown();
+				assignedPartitionSet.removeAll(partitions);
+			}
+
+			@Override
+			public void onPartitionsRevokedAfterCommit(Consumer<?, ?> consumer, Collection<TopicPartition> partitions) {
+				consumerRef.set(consumer);
+				revokedAfterCommitLatch.countDown();
+				assignedPartitionSet.removeAll(partitions);
+			}
+
+			@Override
+			public void onPartitionsAssigned(Consumer<?, ?> consumer, Collection<TopicPartition> partitions) {
+				consumerRef.set(consumer);
+				assignedLatch.countDown();
+				assignedPartitionSet.addAll(partitions);
+			}
+		};
+
+		ContainerProperties containerProps = new ContainerProperties("rebalance.test1");
+		containerProps.setGroupId(groupId);
+		containerProps.setConsumerRebalanceListener(listener);
+		containerProps.setMessageListener((MessageListener<String, String>) (ConsumerRecord<String, String> record) -> {
+		});
+
+		KafkaMessageListenerContainer<String, String> container1 = new KafkaMessageListenerContainer<>(consumerFactory, containerProps);
+		container1.start();
+
+		Thread.sleep(1000); // Wait for initial assignment
+
+		KafkaMessageListenerContainer<String, String> container2 = new KafkaMessageListenerContainer<>(consumerFactory, containerProps);
+		container2.start();
+
+		assertThat(revokedBeforeCommitLatch.await(10, TimeUnit.SECONDS)).isTrue();
+		assertThat(revokedAfterCommitLatch.await(10, TimeUnit.SECONDS)).isTrue();
+		assertThat(assignedLatch.await(10, TimeUnit.SECONDS)).isTrue();
+
+		assertThat(consumerRef.get()).isNotNull();
+		assertThat(consumerRef.get()).isInstanceOf(Consumer.class);
+
+		// In both protocols (new vs legacy), we expect the assignments to contain 6 partitions.
+		// The way they get assigned are dictated by the Kafka server and beyond the scope of this test.
+		// What we are mostly interested is assignment works properly in both protocols.
+		// The new protocol may take several incremental rebalances to attain the full assignments - thus waiting for a
+		// longer duration.
+		Awaitility.await().timeout(Duration.ofSeconds(30))
+				.untilAsserted(() -> assertThat(assignedPartitionSet.size() == 6).isTrue());
+
+		container1.stop();
+		container2.stop();
+	}
+
+	@Configuration
+	@EnableKafka
```

Comment:
Looks like Spring Kafka infrastructure is out of use in this test suite.
More over, I think this test environment (according to test methods body) has nothing to do with Spring at all...

Reply:
I'm not sure what you mean. Although this feature does not require any major changes in Spring Kafka, since it is a major feature in 4.0.0 Kafka, I think we need to verify that it works seamlessly from our side. It's more like a verification that there is no regression in Spring Kafka, regardless of the rebalance protocol used. If any issues are reported in this area, we can potentially use this test as a springboard to debug.  With that said, we can also convert this into a sample app in our examples, and then remove the test.

Reply:
I mean what is a purpose of `@EnableKafka` over here if there is no any `@KafkaListener`?
Why do we need a `@Configuration` at all if Kafka configs can be created as not beans?

therefore my general question: what does Spring do in this test suite at all?

Reply:
Ah ok. got it. I removed the use of `@EnableKafka`, but left the Spring `@Configuration` for the convenience of injecting `EmbeddedKafka`. I think its ok to use it like that for the test, but if you prefer strongly, feel free to rework before the merge. I can also take a look at that later today.